Core/include/Acts/EventData/Measurement.hpp

```diff
@@ -9,387 +9,244 @@
 #pragma once
 
 #include "Acts/Definitions/TrackParametrization.hpp"
-#include "Acts/EventData/ParameterSet.hpp"
 #include "Acts/EventData/SourceLinkConcept.hpp"
-#include "Acts/EventData/TrackParameters.hpp"
-#include "Acts/EventData/detail/fittable_type_generator.hpp"
-#include "Acts/Geometry/Volume.hpp"
-#include "Acts/Surfaces/Surface.hpp"
+#include "Acts/EventData/detail/CalculateResiduals.hpp"
+#include "Acts/EventData/detail/ParameterTraits.hpp"
+#include "Acts/EventData/detail/PrintParameters.hpp"
+#include "Acts/Utilities/detail/Subspace.hpp"
 
-#include <memory>
-#include <ostream>
-#include <type_traits>
-#include <utility>
+#include <cstddef>
+#include <iosfwd>
+#include <variant>
 
 namespace Acts {
 
-// forward declarations
-class Surface;
-
-namespace detail {
-/// @brief Deduction of the measuring geometry object based on the used indices
-template <typename T>
-struct ReferenceObject {};
-template <>
-struct ReferenceObject<BoundIndices> {
-  using type = Surface;
-};
-template <>
-struct ReferenceObject<FreeIndices> {
-  using type = Volume;
-};
-}  // namespace detail
-
-/// @brief base class for Measurements
+/// A measurement of a fixed-size subspace of the full parameters.
 ///
-/// This class describes the measurement of track parameters at a certain
-/// Surface/Volume in the TrackingGeometry.
+/// @tparam source_link_t Source link type to connect to the detector readout
+/// @tparam indices_t Parameter index type, determines the full parameter space
+/// @tparam kSize Size of the parameter subset.
 ///
-/// If the measurement is in local parameters and will not provide localToGlobal
-/// information. It is thus free from any Context.
+/// The measurement is templated on the measurement subspace size and its
+/// content (indices) is defined at runtime. In previous implementations,
+/// the measurement was templated directly on the subspace indices. This lead to
+/// a combinatorial explosion and severe penalty during compilation since all
+/// possible combinations need to be considered for the measurement variant.
+/// However, for the computations only the size of the measurement subspace is
+/// relevant.
```

Comment:
I would remove the mention of previous implementations, I suppose we should keep up with documenting the code as is and not as it was?

Reply:
Ok, I will update the documentation in #555.

Reply:
Merci.